### PR TITLE
Move compatiblity-related code into a separate `compat` module

### PR DESCRIPTION
**Pull request by shadeMe**
There are multiple places in the codebase where we attempt to import libraries inside `try...expect` clauses to safely determine their availability. This PR is primarily meant to replace the individual imports with code that calls into a central `compat` module that maintains the imports for all optional libraries.

Additional changes:
* Removal of the `has_cupy` exports from the `cupy_ops`~, `backends` and `api` modules (these were not explicitly exposed to the public API but used internally)~.
* Handle the deprecation of `cupy.fromDlpack` in `cupy >= 10.0.0`.
* ~`util.set_active_gpu` returns `None` when `cupy`/GPU support is unavailable. Prior to this change, the function could potentially raise an unhandled exception if the `cupy` library was not installed.~ `util.set_active_gpu` now raises an exception if no GPU support is detected.

### CODE REVIEW

**Comment by shadeMe** [comment]
@explosion-bot please test_gpu

File: website/docs/api-util.md [hunk 1]
```diff
@@ -162,10 +163,10 @@ Convert a PyTorch tensor to a `numpy` or `cupy` tensor.
 Convert a `numpy` or `cupy` tensor to a TensorFlow tensor.
 
 | Argument        | Type                       | Description                                           |
-| --------------- | -------------------------- | ----------------------------------------------------- |
+| --------------- | -------------------------- | ----------------------------------------------------- | --- |
 | `xp_tensor`     | <tt>ArrayXd</tt>           | The tensor to convert.                                |
 | `requires_grad` | <tt>bool</tt>              | Whether to backpropagate through the variable.        |
-| `as_variable`   | <tt>bool</tt>              | Convert the result to a `tensorflow.Variable` object. |  |
+| `as_variable`   | <tt>bool</tt>              | Convert the result to a `tensorflow.Variable` object. |     |
```
Comment (svlandeg) [inline website/docs/api-util.md:169]: It looks like the original file had a stray `|` you'll want to remove, as well as the newly added ` --- |` bit in the separator

File: thinc/api.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
 from .util import to_categorical, get_width, get_array_module, to_numpy
 from .util import torch2xp, xp2torch, tensorflow2xp, xp2tensorflow, mxnet2xp, xp2mxnet
 from .backends import get_ops, set_current_ops, get_current_ops, use_ops
-from .backends import Ops, CupyOps, NumpyOps, has_cupy, set_gpu_allocator
+from .backends import Ops, CupyOps, NumpyOps, set_gpu_allocator
```
[the remark this answers is not in the document]
Comment (shadeMe) [reply thinc/api.py:20]: The primary motivation is to reduce the surface area of the public API and have a single source of truth when it comes to presence flags. I can definitely sympathize with your apprehension about [breaking someone's code](https://xkcd.com/1172/) (especially in the case of Python, where literally everything is observable to the end user), but I don't think we are breaking any API guarantees by removing it.

[the remark this answers is not in the document]
Comment (adrianeboyd) [reply thinc/api.py:20]: I agree with Sofie that we should keep this. We usually consider everything that doesn't start with `_` to be part of the public API even if it isn't documented, and for this in particular there is very little cost to keeping it.

File: thinc/backends/_custom_kernels.py [hunk 3]
```diff
@@ -55,7 +51,7 @@
     cupy.RawModule(
         code=KERNELS_SRC, options=("--std=c++11",), name_expressions=KERNELS_LIST
     )
-    if cupy is not None
+    if has_cupy_gpu
```
Comment (danieldk) [inline thinc/backends/_custom_kernels.py:54]: `has_cupy` or `has_cupy_gpu` (I am not sure)? I think the kernels are only compiled at `__call__` time.

File: thinc/backends/__init__.py [hunk 4]
```diff
@@ -174,5 +175,4 @@ def _create_thread_local(
     "Ops",
     "CupyOps",
     "NumpyOps",
-    "has_cupy",
```
Comment (danieldk) [inline thinc/backends/__init__.py:177]: We should also keep this for backward compat.

File: examples/00_intro_to_thinc.ipynb [hunk 5]
```diff
@@ -44,9 +44,9 @@
    },
    "outputs": [],
    "source": [
-    "import thinc.util\n",
+    "import thinc.compat\n",
```
Comment (adrianeboyd) [inline examples/00_intro_to_thinc.ipynb:47]: This should remain backwards-compatible, so let's (make sure that's the case and) undo these changes to the examples.

File: thinc/backends/_custom_kernels.py (hunk 3 above)
Comment (shadeMe) [reply thinc/backends/_custom_kernels.py:54]: We need `has_cupy_gpu` here as `_get_kernel` gets called during module init, which in turn calls into `cupy`. The latter call will raise an exception if there are no visible GPU devices (even if `cupy` is installed, i.e, `has_cupy` returns `True`).

File: thinc/util.py [hunk 6]
```diff
@@ -172,19 +123,18 @@ def to_numpy(data):  # pragma: no cover
...
-    try:
-        import torch
 
+    if has_torch_gpu:
         torch.cuda.set_device(gpu_id)
         torch.set_default_tensor_type("torch.cuda.FloatTensor")
-    except ImportError:
-        pass
+
     return device
```
Comment (svlandeg) [inline thinc/util.py:138]: I worry that this might be slightly breaking behaviour, too. So before we would throw an error if you'd request to set the active GPU but that failed, and now we'd just silently move on?

Comment (shadeMe) [reply thinc/util.py:138]: I guess it depends on how we want to view the original behaviour. Previously, we were importing `cupy` without the necessary sanity checks (like we do elsewhere), opening up the possibility of an unhandled exception. This wasn't specified in the function's documentation (and the "if available" phrase alludes to some amount of exception safety, I think), so I interpreted this lack of a sanity check as a bug.

That said, I now see that the  "if available" phrase could parsed as an adjunct to _just_ `torch`, in which case it would in fact make sense to not modify the function behaviour per-se but to explicitly raise an exception before we call into `cupy`; the `cupy` call will eventually throw under such circumstances anyway). I take that this would be preferable?

Comment (svlandeg) [reply thinc/util.py:138]: I read it as the "if available" being applicable only to the "torch" bit, but it's definitely ambiguous. 

Raising an exception would make sense to me. The docs provide little additional clarification (let's add some) but the current typing to `cupy.cuda.Device` (without `Optional`) signals to me that we'd always like this to work, and if we can't get a proper device, the function should raise an exception.

**Review by svlandeg — COMMENTED** [review]
This looks good to merge to me - just needs to target `master` instead and fix any potential merge conflicts ;-)

**Comment by shadeMe** [comment]
No conflicts! Such a good feeling :grin:

**Comment by svlandeg** [comment]
Just a flaky test. Let's rerun it and see what happens ;-)
